docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,256 @@
+To allow this, settings extensions will be capable of registering themselves as a *validator* for any other setting on the host, including settings that they do not own.
+When handling write requests for multiple settings in a single transaction, the `apiserver` will first gather the resulting writes that would occur under each settings extension involved in the writes.
+Then the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.
+Any one settings extension has the power to halt the transaction by returning a non-zero exit-code on validation.
+
+Validations must specify the setting which triggers the validation in their configuration file:
+
+```toml
+# auditor.toml -- setting extension config file for `settings.auditor`
+[extension.validates]
+network = "v1"
+```
+
+#### Datastore Layout
+Once the `apiserver` has a target set of data to write for a given `set` request and it has been validated by all required validators, it must persist the new settings to the datastore.
+Much like the current datastore, stored settings are stored to the filesystem via a particular pattern.
+
+```bash
+datastore/
+└── motd
+    ├── v1
+    │   └── motd.json
+    └── v2
+        └── motd.json
+```
+Any data written to directories created during a `set` transaction is moved into the datastore.
+Despite only one version of the data being provided, the `apiserver` will populate all versions supported by the settings extension at this time by requesting that the settings extension perform all necessary migrations.
+Failure to populate a version listed as "supported" will result in a failed transaction.
+See the next section about migrations for more details.
+
+### Extension Versioning and Migrations
+Bottlerocket's settings data model changes frequently: sometimes to support new features, or sometimes to correct mistaken model shapes.
+In order to support changes to the settings model, settings must be capable of exposing versioning information about the settings extension, as well as providing faculties for migrating settings data to new versions.
+Our versioning scheme will take heavy influence from Kubernetes' scheme for [versioning CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/).
+Settings extension versioning information is surfaced through the extension config file:
+
+```toml
+# hostname.toml -- setting extension config file for settings.hostname
+[extension]
+supported-versions = [
+     "v1",
+     "v2"
+]
+default-version = "v2"
+```
+
+Any setting can be written or read at any of the `supported-versions` listed by the extension.
+Unversioned requests to read or write this setting are assumed to be the `default-version` and should only take place when a user makes a request using the CLI -- all automated process *should* refer to an explicit version.
```

Comment:
Unversioned requests will also be used when handling user-data, right?

Reply:
Yes!